src/net/server/mod.rs

```diff
@@ -102,53 +131,54 @@ impl MpcHelperServer {
         #[cfg(not(test))]
         const BIND_ADDRESS: Ipv4Addr = Ipv4Addr::UNSPECIFIED;
 
-        let svc = self
-            .router()
-            .layer(
-                TraceLayer::new_for_http()
-                    .make_span_with(move |_request: &hyper::Request<hyper::Body>| {
-                        tracing.make_span()
-                    })
-                    .on_request(|request: &hyper::Request<hyper::Body>, _: &Span| {
-                        increment_counter!(RequestProtocolVersion::from(request.version()));
-                        increment_counter!(REQUESTS_RECEIVED);
-                    }),
-            )
-            .into_make_service();
+        let svc = self.router().layer(
+            TraceLayer::new_for_http()
+                .make_span_with(move |_request: &hyper::Request<hyper::Body>| tracing.make_span())
+                .on_request(|request: &hyper::Request<hyper::Body>, _: &Span| {
+                    increment_counter!(RequestProtocolVersion::from(request.version()));
+                    increment_counter!(REQUESTS_RECEIVED);
+                }),
+        );
         let handle = Handle::new();
 
         let task_handle = match (self.config.disable_https, listener) {
             (true, Some(listener)) => {
+                let svc = svc
+                    .layer(layer_fn(SetClientIdentityFromHeader::new))
```

Comment:
would it be a valid configuration with HTTPS enabled and `SetClientIdentityFromHeader` used instead of `ClientCertRecognizingAcceptor`? If so, is there any reasonable way we can protect against it?

Reply:
> would it be a valid configuration with HTTPS enabled and SetClientIdentityFromHeader used instead of ClientCertRecognizingAcceptor? If so, is there any reasonable way we can protect against it?

The current code doesn't allow enable/disable of SetClientIdentityFromHeader middleware other than via the HTTPS disable. So it's not possible to do something bad via configuration.

Whether we want to protect against (intentional or not) changes in behavior as code evolves is also a valid question. One strategy is to remove it from production builds entirely with conditional compilation -- like what `weak-field` does for `Fp31`. This is probably what I would pick, but I'm not sure if we want to add more build variants to the mix right now.

Another option might be to add some sort of interlock with the configuration. e.g. if `network.toml` is marked as production, then we disallow activating insecure modes. This could be done as a redundant check in `SetClientIdentityFromHeader::new`, which conveniently already has a copy of NetworkConfig.

(Side note, I'll also remove `pub(super)` from `SetClientIdentityFromHeader` -- it is left over from an earlier version where I tried to make `HTTP_CLIENT_ID_HEADER` an associated constant.)

Reply:
Yea, I was thinking along the same lines - if there is only one boolean check that prevents us from getting to use  `SetClientIdentityFromHeader` with HTTPS (in production), more safety nets would be good to have. maybe the interlock is everything what we need for now.

Reply:
I am going to merge this so that we can resolve the conflicts with #654, but I opened #660 to add the production interlock.